### DISCUSSION TITLE
avalonia.js 404

### DISCUSSION THREAD

**Discussion by githubxiaoa** [AvaloniaUI/Avalonia#16467] [Q&A]
### Describe the bug

avalonia.js 404
![微信截图_20240724170111](https://github.com/user-attachments/assets/0e1c5b5d-5ab5-4792-a67a-9c0b23911b66)

### To Reproduce

not run

### Avalonia version

11.1.0

### OS

Windows, WebAssembly

**Comment by maxkatz6** [comment]
Try to recreate browser project from the avalonia.xplat template.

**Comment by githubxiaoa** [comment]
I have created a new project without making any changes, but it is still unsuccessful

**Comment by msneijders** [comment]
Using template v1.14 it works running from visual studio (it finds and reads an avalonia.js file) -- but when I deploy (testing with a local IIS), I cannot find the avalonia.js file in `_framework` folder. I copied the content from avalonia.js that is generated while debugging from visual studio from within the browsers DevTools (because I cannot find it in wwwroot/_framework), and put it in the _framework folder of the deployment; then it runs.

**Comment by maxkatz6** [comment]
> I have created a new project without making any changes, but it is still unsuccessful

It doesn't give much information on how you run it.
New project templates work just fine here, via `dotnet run` and `dotnet publish` (which are the most important).
If any IDE doesn't handle running projects - report the issue to them.

> testing with a local IIS

Can't really help with IIS. Don't see reasons to use it also.
For local testing of published packages, I recommend [dotnet serve](https://github.com/natemcmaster/dotnet-serve) tool. For production there are various static website cloud providers as well, or nginx should work too just fine.

> but when I deploy

Do you deploy with `dotnet publish` btw? Any other way won't work.

**Reply by msneijders** [reply]
> Do you deploy with `dotnet publish` btw? Any other way won't work.

:flushed: I used 'publish' from within visual studio -- naively thinking that should work, it does keep saying 'publish succeeded'. But simply using `dotnet publish` I do get a full subfolder '_framework', containing the `avalonia.js`.

**Comment by ztl8702** [comment]
I noticed that the `avalonia.js` StaticWebAsset item is created with `SourceType=Package`:

https://github.com/AvaloniaUI/Avalonia/blob/0bd90d802a208e4d4247d188a3fadbe129087f2e/src/Browser/Avalonia.Browser/build/Microsoft.AspNetCore.StaticWebAssets.props#L9

However, when I inspect the MSBuild binlog for a project that _consumes_ `Avalonia.Browser` **transitively**, I noticed that `dotnet.js` ends up having `SourceType=Computed`. (I don't know how/why, since StaticWebAssetSdk is not well documented. But the key here is that `avalonia.js` has different metadata than `dotnet.js`):

<details>
<img width="1006" alt="Screenshot 2024-10-15 at 13 02 25" src="https://github.com/user-attachments/assets/82c862ad-6d5d-4666-84f6-fe857c4be014">
</details>

And hence it is skipped by the `CopyStaticWebAssetsToOutputDirectory` target. ([SDK source](https://github.com/dotnet/sdk/blob/ef306ab4fff1ef29d6c1315c3653f84e3d649a14/src/StaticWebAssetsSdk/Tasks/CollectStaticWebAssetsToCopy.cs#L29-L33))

<img width="1034" alt="Screenshot 2024-10-15 at 13 03 50" src="https://github.com/user-attachments/assets/2435265a-a674-49b4-80d6-eab2d5a5c1a6">


I am using .NET SDK `9.0.100-preview.3.24204.13`. The project targets .NET 8.0. 

MSBuild Binlog of the _consuming project_: 
[build.binlog.zip](https://github.com/user-attachments/files/17371169/build.binlog.zip)

**Reply by ztl8702** [reply]
Note that for me `dotnet publish` does output `avalonia.js` correctly. I am specifically talking about the following use-case:

* (1) Main Project (AspNetCore web server app)
   * references a (2) `Microsoft.NET.Sdk.WebAssembly` project (with `ReferenceOutputAssembly="false" SkipGetTargetFrameworkProperties="true" OutputItemType="Content"`, to consume its static content), 
      * which references `Avalonia.Browser`
      
directly publishing (2) will work. But building (1) does _not_ correctly copy `avalonia.js` to (1), whereas `dotnet.js` will be copied to (1).

Adding `Avalonia.Browser` directly to (1) does not fix the issue either.

**Reply by ITDancer13** [reply]
I had the same issue. Running `dotnet workload install wasm-tools` and rebuilding everything fixed it. The default application of the template works.

**Reply by suprak** [reply]
> I had the same issue. Running `dotnet workload install wasm-tools` and rebuilding everything fixed it. The default application of the template works.

After installing wasm tools and even restarting VS, launching from VS did not work for me.
Running `dotnet run` from the Browser project folder did work.

**Reply by cbordeman** [reply]
None of this works.  The js file is still not present.

**Reply by Qzername** [reply]
None of this works for me too.

Is there a solution or something to resolve this problem?

**Reply by cbordeman** [reply]
@suprak Don't use the VS templates, they're totally broken and haven't been updated in a while.  Create your projects using the dotnet command.

**Reply by Qzername** [reply]
I did some digging, and as the @cbordeman said, VS templates are broken for some reason. 

When you create a project with dotnet command everything works properly.

**Reply by luoczeng** [reply]
I wish I had seen this discussion earlier. I had the same issue and had to manually copy avalonia.js.
See #18101

**Reply by ZzZombo** [reply]
For me I had to manually trigger build output clean up before it started to publish properly.
